File: LeetCode_Solution/014. Longest Common Prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) {
            return "";
        }
        
        // 以数组中第一个字符串为基准，后续字符串都和第一个比较即可
        String p = strs[0];
        
        for (int i = 1; i < strs.length; i++) {
            int j = 0;
            
            // 判断每个字符串中的字母的指针j，必须小于当前比较的字符串和首个字符串的小者
            // 然后判断当前字符串的同位置的字符和基准字符是否相等
            while (j < strs[i].length() && j < p.length() && strs[i].charAt(j) == p.charAt(j)) {
                j++;
            }
            
            // 如果没有字符串相同，返回空字符串
            if (j == 0) {
                return "";
            }
            
            // 有字符串相同，以首字符串为基准，返回相同的最大子串
            p = p.substring(0,j);
        }
        
        return p;
    }
} 

File: java_solution/014. Longest Common Prefix.py
class Solution:
    def longestCommonPrefix1(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        
        result = ""
        
        i = 0
        while True:
            try:
                sets = set(str[i] for str in strs)
                if len(sets) == 1:
                    result += sets.pop()
                    i += 1
                else:
                    break
                    
            except Exception as e:
                break
        
        return result
    
    
    
    def longestCommonPrefix2(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        
        # in case of empty input list
        if not strs:
            return ""
        
        for i in range(len(strs[0])):
            for str in strs[1:]:
                if i >= len(str) or str[i] != strs[0][i]:
                    return strs[0][:i]
        
        # in case of the first string in the input list is ""
        return strs[0]
    
    
    
